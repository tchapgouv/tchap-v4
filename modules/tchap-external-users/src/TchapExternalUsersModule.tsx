/*
Copyright 2022 New Vector Ltd. t/a Element

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { RuntimeModule } from "@matrix-org/react-sdk-module-api/lib/RuntimeModule";
import { ModuleApi } from "@matrix-org/react-sdk-module-api/lib/ModuleApi";
import { CantAddDialog  } from "./components/CantAddDialog";
import { AddRoomListener, ExternalUsersLifecycle } from "@matrix-org/react-sdk-module-api/lib/tchap/lifecycles/ExternalUsers";
import { DialogProps } from "@matrix-org/react-sdk-module-api/lib/components/DialogContent";

import React from "react";

export default class TchapExternalUsersModule extends RuntimeModule {
    private domain: String;

    public constructor(moduleApi: ModuleApi) {
        super(moduleApi);

        // const rawConfig = this.moduleApi.getConfigValue(
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_NAMESPACE,
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_KEY,
        //   );
      
        //   assertValidGuestModuleConfig(rawConfig);
        //   this.config = rawConfig;

        this.on(ExternalUsersLifecycle.AddRoom, this.onAddRoom);
    }

    protected onAddRoom: AddRoomListener = async () => {
        await this.moduleApi.openDialog<null, DialogProps, CantAddDialog>("test", (props, ref) => <CantAddDialog ref={ref} {...props} />);
    };
}
