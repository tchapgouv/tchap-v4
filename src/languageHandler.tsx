/*
Copyright 2024 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import {
    IVariables,
    TranslatedString,
    TranslationKey as ReactTranslationKey,
    // eslint-disable-next-line camelcase
    _t as react_t,
    // eslint-disable-next-line camelcase
    _td as react_td,
    // eslint-disable-next-line camelcase
    _tDom as react_tDom,
    Tags,
    UserFriendlyError as ReactUserFriendlyError,
    ErrorOptions,
} from "matrix-react-sdk/src/languageHandler";
import { Leaves } from "matrix-react-sdk/src/@types/common";

import type ReactEN from "matrix-react-sdk/src/i18n/strings/en_EN.json";
import type EN from "./i18n/strings/en_EN.json";

/**
 * This module wraps languageHandler in the matrix-react-sdk and adds type casts to include translations
 * which we know will be injected by webpack.
 */

export type TranslationKey = Leaves<typeof EN & typeof ReactEN, "|", string | { other: string }, 4>;

export class UserFriendlyError extends ReactUserFriendlyError {
    public constructor(message: TranslationKey, substitutionVariablesAndCause?: IVariables & ErrorOptions) {
        super(message as ReactTranslationKey, substitutionVariablesAndCause);
    }
}

export function _td(s: TranslationKey): TranslationKey {
    return react_td(s as ReactTranslationKey);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _t(text: TranslationKey, variables?: IVariables): string;
export function _t(text: TranslationKey, variables: IVariables | undefined, tags: Tags): React.ReactNode;
export function _t(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_t(text as ReactTranslationKey, variables, tags!);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _tDom(text: TranslationKey, variables?: IVariables): TranslatedString;
export function _tDom(text: TranslationKey, variables: IVariables, tags: Tags): React.ReactNode;
export function _tDom(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_tDom(text as ReactTranslationKey, variables!, tags!);
}
