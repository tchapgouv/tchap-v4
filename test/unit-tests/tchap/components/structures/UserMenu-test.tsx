import React from "react";
import { render, screen } from "@testing-library/react";
import { MatrixClient } from "matrix-js-sdk/src/matrix";

import UnwrappedUserMenu from "~matrix-react-sdk/src/components/structures/UserMenu";
import { TestSdkContext } from "~matrix-react-sdk/test/TestSdkContext";
import { stubClient, wrapInSdkContext } from "~matrix-react-sdk/test/test-utils";

describe("<UserMenu>", () => {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    let client: MatrixClient;
    let sdkContext: TestSdkContext;

    beforeEach(() => {
        client = stubClient();
        sdkContext = new TestSdkContext();
    });

    describe("<UserMenu> UI", () => {
        beforeEach(() => {
            const UserMenu = wrapInSdkContext(UnwrappedUserMenu, sdkContext);
            render(<UserMenu isPanelCollapsed={false} />);
        });

        // If this snapshot change, you should consider to check what change on the UI side from element
        // it"s a good indicator to see if this could introduce some regression on our code
        it("should render as expected", async () => {
            // open the user menu
            screen.getByRole("button", { name: "a11y" }).click();

            const menu = screen.getByRole("menu");
            expect(menu).toMatchSnapshot();
        });

        // you can also add some specific ui check here
        // ...
    });

    // Here you can add some business logic check
    describe("<UserMenu> faq", () => {
        beforeEach(() => {
            const UserMenu = wrapInSdkContext(UnwrappedUserMenu, sdkContext);
            render(<UserMenu isPanelCollapsed={true} />);
        });

        it("should open the faq when clicking on the faq button", () => {
            global.open = jest.fn();
            // open the user menu
            screen.getByRole("button", { name: "a11y" }).click();
            // click on the faq
            screen.getByRole("menuitem", { name: "common" }).click();
            expect(global.open).toHaveBeenCalledTimes(1);
            expect(global.open).toHaveBeenCalledWith("https://www.tchap.gouv.fr/faq", "_blank");
        });
    });

    // Here are the tests for checking the props of the component
    describe("<UserMenu> props", () => {});
});
