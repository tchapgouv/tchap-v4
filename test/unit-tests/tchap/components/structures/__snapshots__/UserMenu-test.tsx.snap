// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserMenu> <UserMenu> UI should render as expected 1`] = `
<div
  class="mx_ContextualMenu"
  role="menu"
>
  <ul
    class="mx_IconizedContextMenu mx_UserMenu_contextMenu"
    role="none"
  >
    <div
      class="mx_UserMenu_contextMenu_header"
    >
      <div
        class="mx_UserMenu_contextMenu_name"
      >
        <span
          class="mx_UserMenu_contextMenu_displayName"
        />
        <span
          class="mx_UserMenu_contextMenu_userId"
        >
          @userId:matrix.org
        </span>
      </div>
      <div
        aria-label="Switch to dark mode"
        class="mx_AccessibleButton mx_UserMenu_contextMenu_themeButton"
        role="button"
        tabindex="-1"
      >
        <img
          alt=""
          role="presentation"
          src="image-file-stub"
          width="16"
        />
      </div>
    </div>
    <div
      class="mx_IconizedContextMenu_optionList mx_IconizedContextMenu_optionList_notFirst"
    >
      <li
        aria-label="Link new device"
        class="mx_AccessibleButton mx_IconizedContextMenu_item"
        role="menuitem"
        tabindex="0"
      >
        <span
          class="mx_IconizedContextMenu_icon mx_UserMenu_iconQr"
        />
        <span
          class="mx_IconizedContextMenu_label"
        >
          Link new device
        </span>
      </li>
      <li
        aria-label="Notifications"
        class="mx_AccessibleButton mx_IconizedContextMenu_item"
        role="menuitem"
        tabindex="-1"
      >
        <span
          class="mx_IconizedContextMenu_icon mx_UserMenu_iconBell"
        />
        <span
          class="mx_IconizedContextMenu_label"
        >
          Notifications
        </span>
      </li>
      <li
        aria-label="Security & Privacy"
        class="mx_AccessibleButton mx_IconizedContextMenu_item"
        role="menuitem"
        tabindex="-1"
      >
        <span
          class="mx_IconizedContextMenu_icon mx_UserMenu_iconLock"
        />
        <span
          class="mx_IconizedContextMenu_label"
        >
          Security & Privacy
        </span>
      </li>
      <li
        aria-label="All settings"
        class="mx_AccessibleButton mx_IconizedContextMenu_item"
        role="menuitem"
        tabindex="-1"
      >
        <span
          class="mx_IconizedContextMenu_icon mx_UserMenu_iconSettings"
        />
        <span
          class="mx_IconizedContextMenu_label"
        >
          All settings
        </span>
      </li>
      <li
        aria-label="Help"
        class="mx_AccessibleButton mx_IconizedContextMenu_item"
        role="menuitem"
        tabindex="-1"
      >
        <span
          class="mx_IconizedContextMenu_icon mx_UserMenu_iconInfo"
        />
        <span
          class="mx_IconizedContextMenu_label"
        >
          Help
        </span>
      </li>
      <li
        aria-label="Sign out"
        class="mx_AccessibleButton mx_IconizedContextMenu_option_red mx_IconizedContextMenu_item"
        role="menuitem"
        tabindex="-1"
      >
        <span
          class="mx_IconizedContextMenu_icon mx_UserMenu_iconSignOut"
        />
        <span
          class="mx_IconizedContextMenu_label"
        >
          Sign out
        </span>
      </li>
    </div>
  </ul>
</div>
`;
